src/native/clr/host/bridge-processing.cc

```diff
@@ -0,0 +1,307 @@
+	if (scc.Count > 0) {
+		abort_unless (scc.Contexts [0] != nullptr, "SCC must have at least one context");
+		abort_unless (scc.Contexts [0]->control_block != nullptr, "SCC must have at least one context with valid control block");
+		return { .is_temporary_peer = false, .context = scc.Contexts [0] };
+	} else {
+		const auto temporary_peer = temporary_peers.find (scc_index);
+		abort_unless (temporary_peer != temporary_peers.end(), "Temporary peer must be found in the map");
+		return { .is_temporary_peer = true, .temporary_peer = temporary_peer->second };
+	}
+}
+
+void BridgeProcessing::add_circular_references (const StronglyConnectedComponent &scc) noexcept
+{
+	abort_unless (scc.Count > 1, "SCC must have at least two items to add inner references");
+
+	JniObjectReferenceControlBlock *prev = scc.Contexts [scc.Count - 1]->control_block;
+
+	for (size_t j = 1; j < scc.Count; j++) {
+		JniObjectReferenceControlBlock *next = scc.Contexts [j]->control_block;
+		bool reference_added = add_reference (prev->handle, next->handle);
+	
+		// TODO this doesn't seem to be handled in the Mono code but if we don't handle this case, then the component _might not be_ strongly connected on the Java side.
+		// What is the case when this fails though? All bridge objects should have the `monodroidAddReference` method, so this is unexpected.
+		abort_unless (reference_added, [this, &prev, &next] {
+			jclass prev_java_class = env->GetObjectClass (prev->handle);
+			const char *prev_class_name = Host::get_java_class_name_for_TypeManager (prev_java_class);
+
+			jclass next_java_class = env->GetObjectClass (next->handle);
+			const char *next_class_name = Host::get_java_class_name_for_TypeManager (next_java_class);
+
+			return detail::_format_message (
+				"Failed to add reference between objects in a strongly connected component: %s -> %s.",
+				prev_class_name,
+				next_class_name);
+		});
+	
+		prev->refs_added = 1;
+		prev = next;
+	}
+}
+
+void BridgeProcessing::add_cross_reference (size_t source_index, size_t dest_index) noexcept
+{
+	CrossReferenceTarget from = select_cross_reference_target (source_index);
+	CrossReferenceTarget to = select_cross_reference_target (dest_index);
+
+	if (add_reference (from.get_handle(), to.get_handle())) {
+		from.mark_refs_added_if_needed ();
+	}
+}
+
+bool BridgeProcessing::add_reference (jobject from, jobject to) noexcept
+{
+	jclass java_class = env->GetObjectClass (from);
+	jmethodID add_method_id = env->GetMethodID (java_class, "monodroidAddReference", "(Ljava/lang/Object;)V");
+	env->DeleteLocalRef (java_class);
+
+	if (add_method_id != nullptr) {
+		env->CallVoidMethod (from, add_method_id, to);
+		return true;
+	} else {
+		env->ExceptionClear ();
+		return false;
+	}
+}
+
+void BridgeProcessing::clear_references_if_needed (JniObjectReferenceControlBlock *control_block) noexcept
+{
+	abort_if_invalid_pointer_argument (control_block, "control_block");
```

Comment:
`control_block` is validated at call site. Since this is a private function, add a comment above it describing the precondition and remove this line.